tests/test/storage/test_file_descriptor.cpp

```diff
@@ -238,4 +227,121 @@ TEST_CASE("Check stat and read shared file", "[storage]")
+
+    // Get the full directory listing using stdlib
+    DIR* dir = opendir(dirPath.c_str());
+    std::vector<std::string> expectedList;
+    struct dirent* ent;
+    while ((ent = readdir(dir)) != nullptr) {
+        expectedList.push_back(ent->d_name);
+    }
+    closedir(dir);
+
+    // Check there are actually some files in the directory
+    REQUIRE(expectedList.size() > 50);
+
+    // Open the directory
+    int dirFd = fs.openFileDescriptor(
+      DEFAULT_ROOT_FD, wasmPath, 0, 0, 0, __WASI_O_DIRECTORY, 0);
+    REQUIRE(dirFd > 0);
+
+    storage::FileDescriptor& fileDesc = fs.getFileDescriptor(dirFd);
+
+    SECTION("Iterator")
+    {
+        REQUIRE(fileDesc.iterStarted() == false);
+        REQUIRE(fileDesc.iterFinished() == false);
+
+        // Make sure first few items are the same
+        int step = 3;
+        for (int i = 0; i < step; i++) {
+            storage::DirEnt ent = fileDesc.iterNext();
+            REQUIRE(ent.next == i + 1);
+            REQUIRE(ent.path == expectedList.at(i));
+        }
+
+        REQUIRE(fileDesc.iterStarted() == true);
+        REQUIRE(fileDesc.iterFinished() == false);
+
+        // Go back one in the iterator and check we get the relevant entry
+        fileDesc.iterBack();
+        storage::DirEnt backEnt = fileDesc.iterNext();
+        REQUIRE(backEnt.path == expectedList.at(step - 1));
+
+        // Reset, and walk through the whole iterator to check the values match
+        fileDesc.iterReset();
+        REQUIRE(fileDesc.iterStarted() == false);
+        REQUIRE(fileDesc.iterFinished() == false);
+
+        std::vector<std::string> actualList;
+        while (!fileDesc.iterFinished()) {
+            actualList.push_back(fileDesc.iterNext().path);
+        }
+
+        REQUIRE(actualList == expectedList);
+        REQUIRE(fileDesc.iterStarted() == true);
+        REQUIRE(fileDesc.iterFinished() == true);
+    }
+
+    SECTION("WASI dirent buffer")
+    {
+        // Get the first three entries
+        storage::DirEnt entA = fileDesc.iterNext();
+        storage::DirEnt entB = fileDesc.iterNext();
+        storage::DirEnt entC = fileDesc.iterNext();
+
+        // Work out how long a buffer needs to be
+        size_t wasiDirentSize = sizeof(__wasi_dirent_t);
+        size_t sizeA = wasiDirentSize + entA.path.size();
+        size_t sizeB = wasiDirentSize + entB.path.size();
+        size_t sizeC = wasiDirentSize + entC.path.size();
+
+        // Reset the iterator
+        fileDesc.iterReset();
+
+        // Make a buffer slightly too small for all of them
+        std::vector<uint8_t> buffer(sizeA + sizeB + sizeC - 10);
+
+        // Copy into this buffer
+        size_t bytesCopied =
+          fileDesc.copyDirentsToWasiBuffer(buffer.data(), buffer.size());
+        REQUIRE(bytesCopied == buffer.size());
+
+        // Check contents
+        checkWasiDirentInBuffer(buffer.data(), entA);
+        checkWasiDirentInBuffer(buffer.data() + sizeA, entB);
```

Comment:
Could we actually check that `entC` is _not_ in the buffer?

Reply:
Only part of `entC` would fit, so we'd be checking whether _some_ of `entC` had been written to the buffer (might be a bit fiddly). 

More importantly, it's undefined as to whether the function should or shouldn't write this partial dirent. The libc wrapper will always skip reading this last part of the buffer, unless there's at least a full dirent's worth of bytes left (i.e. it will never read this part of the buffer and it's undefined as to what should be written there).